dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,216 @@
+          mls = M.fromList . fst .
+                foldl (\(ls, lsl) (n, l) ->
+                           let lsl' = lsl + 1 + T.length l in
+                           ((n, (lsl, l)):ls, lsl')) ([], 0) . zip [1..] $ T.lines t
+
+getTagsFromExpr :: Expr Src a -> M.Map PosInFile Tag
+getTagsFromExpr = go (FP 0 0) M.empty
+    where go lpos mts = \case
+              (Let b e) -> go lpos (mts <> parseBinding b) e
+              (Annot e1 e2) -> go lpos (go lpos mts e1) e2
+              (Record mr) -> mts <> tagsFromDhallMap lpos mr
+              (RecordLit mr) -> mts <> tagsFromDhallMap lpos mr
+              (Union mmr) -> mts <> tagsFromDhallMap lpos mmr
+              (Note s e) -> go (srcToPosInFile s) mts e
+              _ -> mts
+          tagsFromDhallMap lpos = foldMapWithKey (\k _ -> M.singleton lpos (Tag "" k))
+
+parseBinding :: Binding Src a -> M.Map PosInFile Tag
+parseBinding b = M.singleton (maybe (FP 0 0) srcToPosInFile (bindingSrc0 b))
+                             (Tag "" (variable b))
+
+srcToPosInFile :: Src -> PosInFile
+srcToPosInFile s = FP line column
+    where ssp = srcStart s
+          line = unPos . sourceLine $ ssp
+          column = unPos . sourceColumn $ ssp
+
+showTags :: Tags -> Text
+showTags (Tags ts) = T.concat . map (uncurry showFileTags) . M.toList $ ts
+
+showFileTags :: FilePath -> M.Map PosInFile Tag -> T.Text
+showFileTags f ts = "\x0c\n" <> T.pack f <> "," <> (showInt . T.length) cs <> "\n" <> cs
+    where cs = T.concat . map (uncurry showPosTag) . M.toList $ ts
+
+showPosTag :: PosInFile -> Tag -> Text
+showPosTag fp tag = def <>"\x7f" <> name <> "\x01" <> showInt line <>
+                    "," <> showInt offset <> "\n"
+    where line = fpLine fp
+          offset = fpOffset fp
+          def = tagDefinition tag
+          name = tagName tag
+
+showInt :: Int -> Text
+showInt = T.pack . show
+
+{-| Generate list of files for given Input
+-}
+inputToFiles
+    :: Bool
+    -- ^ In case true, function will get files from  symbolic links
+    -> [String]
+    -- ^ List of suffixes. In case it contains empty string (""), all files will be returned.
+    --   This parameter works only in case input is path to directory.
+    -> Input
+    -- ^ Where to look fo files. Can be directory name (`.` for example),
+    --   file name or StandardInput. In case of StandardInput will wait for file names from
+    --   STDIN, This way someone can combine tools in bash to send, for example, output from
+    --   find to input of dhall tags.
+    -> IO [ FilePath ]
+    --   List of files.
+inputToFiles _ _ StandardInput = lines <$> getContents
+inputToFiles followSyms suffixes (InputFile path) = go path
+    where go p = do
+                   isD <- doesDirectoryExist p
+#if MIN_VERSION_directory(1,3,0)
+                   isSymLink <- pathIsSymbolicLink p
+#elif MIN_VERSION_directory(1,2,6)
+                   isSymLink <- isSymbolicLink pa
+#else
+                   let isSymLink = False
+#endif
```

Comment:
The inline CPP is makes this pretty hard to read. Can you move this to the `where`-clause below?

Reply:
Done.